pkg/virt-controller/ipamclaims/ipamclaims.go

```diff
@@ -0,0 +1,185 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+
+package ipamclaims
+
+import (
+	"context"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	virtv1 "kubevirt.io/api/core/v1"
+
+	"kubevirt.io/kubevirt/pkg/network/namescheme"
+	"kubevirt.io/kubevirt/pkg/network/vmispec"
+	"kubevirt.io/kubevirt/pkg/virt-controller/ipamclaims/libipam"
+	"kubevirt.io/kubevirt/pkg/virt-controller/network"
+
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+
+	ipamclaims "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1"
+	ipamclaimsclient "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1/apis/clientset/versioned"
+
+	networkclient "kubevirt.io/client-go/generated/network-attachment-definition-client/clientset/versioned"
+)
+
+type IPAMClaimsManager struct {
+	networkClient    networkclient.Interface
+	ipamClaimsClient ipamclaimsclient.Interface
+}
+
+func NewIPAMClaimsManager(networkClient networkclient.Interface, ipamClaimsClient ipamclaimsclient.Interface) *IPAMClaimsManager {
+	return &IPAMClaimsManager{
+		networkClient:    networkClient,
+		ipamClaimsClient: ipamClaimsClient,
+	}
+}
+
+func (m *IPAMClaimsManager) CreateNewPodIPAMClaims(vmi *virtv1.VirtualMachineInstance, ownerRef *v1.OwnerReference) error {
+	if ownerRef == nil {
+		ownerRef = v1.NewControllerRef(vmi, virtv1.VirtualMachineInstanceGroupVersionKind)
+	}
+	nonAbsentNetworks := filterNonAbsentNetworks(vmi.Spec.Domain.Devices.Interfaces, vmi.Spec.Networks)
+	networkToIPAMClaimParams, err := m.GetNetworkToIPAMClaimParams(
+		vmi.Namespace,
+		vmi.Name,
+		vmispec.FilterMultusNonDefaultNetworks(nonAbsentNetworks))
+	if err != nil {
+		return fmt.Errorf("failed composing networkToIPAMClaimName: %w", err)
+	}
+
+	claims := composeIPAMClaims(vmi.Namespace, ownerRef, networkToIPAMClaimParams)
+	if err := m.createNewPodIPAMClaims(vmi.Namespace, claims); err != nil {
+		return fmt.Errorf("failed IPAMClaims creation for VMI %s: %w", vmi.Name, err)
+	}
+
+	return nil
+}
+
+func composeIPAMClaims(namespace string, ownerRef *v1.OwnerReference, networkToIPAMClaimParams map[string]libipam.IPAMClaimParams) []*ipamclaims.IPAMClaim {
+	claims := []*ipamclaims.IPAMClaim{}
+	for netName, ipamClaimParams := range networkToIPAMClaimParams {
+		claims = append(claims, composeIPAMClaim(
+			namespace,
+			*ownerRef,
+			ipamClaimParams,
+			namescheme.GenerateHashedInterfaceName(netName),
+		))
+	}
+
+	return claims
+}
+
+func (m *IPAMClaimsManager) createNewPodIPAMClaims(namespace string, claims []*ipamclaims.IPAMClaim) error {
+	for _, claim := range claims {
+		_, err := m.ipamClaimsClient.K8sV1alpha1().IPAMClaims(namespace).Create(
```

Comment:
IMO it is better to label the IPAMClaims w/ the name of the kubevirt VM when creating.

This way, we can always query the datastore for all claims belonging to a certain VM.

Reply:
> it can wait for hot plug right ?

I don't follow what hot plug has to do with this.

Querying all the claims belonging to a VM is useful for removing the finalizer when a VMI object is removed from the datastore.

Up until that code lands, it is useful for debugging purposes - I assume that's what you mean w/ "manual queries".

> also note we support VMI so in that case we need to label with VMI name

There's no need to differentiate between vmi / vm since they're the same - i.e. vmi name == vm name

> btw maybe it allows to query according owner ref without label at all ?

I don't think so, if you know it's possible, please educate me. AFAIK, you can only filter using either label selectors, or field selectors. And not all fields can be used. Here you can find a list of usable fields for pods in 1.30.1: https://github.com/kubernetes/kubernetes/blob/v1.30.1/pkg/registry/core/pod/strategy.go#L343

We want to use the query eventually from the code. It's a lot simpler to use a label than to use the thing you've pasted.

Reply:
> I don't follow what hot plug has to do with this.

lets add it please when it is needed, i.e when adding finalizers / hot plug (this what i meant)

> I don't think so, if you know it's possible, please educate me.

Just meant meanwhile for debugging, one can do a query like in the first message (with json parsing etc)
not from the code.
Once we need to add the relevant code lets add it.
If possible, lets please freeze a state, we can always add stuff later, otherwise it won't converge.